src/gg20/crypto_tools/paillier/zk/composite_dlog.rs

```diff
@@ -80,51 +101,108 @@ fn jacobi_symbol(a: &BigNumber, p: &BigNumber, q: &BigNumber) -> i32 {
 impl CompositeDLogStmt {
     #[allow(clippy::many_single_char_names, non_snake_case)]
     /// Setup the statement for Composite Dlog proof using the modulus `N = pq`.
-    /// This will generate an asymmetric basis `g` and a witness `s`.
+    /// This will generate an asymmetric basis `g` and a witness `s` such that
+    /// g^(-s) is also asymmetric basis.
+    /// The first statement returned is for `v = g^(-s)` with witness `s`,
+    /// and the second statment returned is for `g = v^(-s^(-1))` with witness `s^(-1)`.
     pub fn setup(
         rng: &mut (impl CryptoRng + RngCore),
         n: &BigNumber,
         p: &BigNumber,
         q: &BigNumber,
-    ) -> (Self, Randomness) {
-        // Sample an asymmetric basis g
-        let g = loop {
+        totient: &BigNumber,
+    ) -> (Self, CompositeDLogWitness, Self, CompositeDLogWitness) {
+        // We sample from S_WITNESS_SIZE - 1 and add a bit at the end to get an invertible element quickly
+        let S_div_2 = BigNumber::one() << (S_WITNESS_SIZE - 1);
+
+        loop {
+            // Sample an asymmetric basis g
             let g = BigNumber::random_with_rng(rng, n);
 
             // g is asymmetric when Jacobi symbol (g | n) = -1
-            if jacobi_symbol(&g, p, q) == -1 {
-                break g;
+            if jacobi_symbol(&g, p, q) != -1 {
+                continue;
+            }
+
+            // Sample s from {0,..,S-1} such that it is in Z*_phi(N)
+            // If p and q are safe primes, then any odd number `s` is invertible with very high probability.
+            // If p and q are not safe primes, then this can occur often, and we need to resample.
+            let (s, s_inv) = loop {
+                let s = Randomness((BigNumber::random_with_rng(rng, &S_div_2) << 1) + 1);
+
+                // Inversion will fail if s is not co-prime to phi(N)
+                match s.0.invert(totient) {
+                    None => {
+                        warn!("cryptographically unreachable (except for unsafe primes): random `s` not in `Z*_phi(n)`. trying again...");
+                    }
+                    Some(x) => {
+                        break (s, Randomness(x));
+                    }
+                };
+            };
+
+            let neg_s = Randomness(-&s.0);
+
+            // v = g^(-s) mod N
+            let v = g.modpow(&neg_s.0, n);
+
+            // Check if v is also asymmetric
+            if jacobi_symbol(&v, p, q) != -1 {
```

Comment:
I think if `s` is invertible `mod phi(N)` and if `g` is asymmetric (quadratic nonresidue), then `v` should always be asymmetric too: if there was `x` such that `x^2 = v = g^(-s) mod N`, then `g = (x^(-s^(-1)))^2 mod N` making `g` a quadratic residue. But the paranoid part of me left this in.

Reply:
So, to be asymmetric `v` would have to be quadratic nonresidue mod only one of `p` or `q`. making it a nonresidue mod `N`. It could be that it's a quadratic nonresidue mod both `p` and `q` perhaps?